nixd/lib/Server/Controller.cpp

```diff
@@ -233,9 +240,9 @@ void Controller::readJSONConfig(lspserver::PathRef File) noexcept {
     std::ifstream In(File.str(), std::ios::in);
     SS << In.rdbuf();
 
-    if (auto NewConfig = parseConfig(SS.str()))
-      updateConfig(std::move(NewConfig.get()));
-    else {
+    if (auto NewConfig = parseConfig(SS.str())) {
+      JSONConfig = std::move(NewConfig.get());
```

Comment:
I do think it is a good idea to separate `JSONConfig` from `Config`, but I think maybe you can write a method like 

```
getXXXField(xxx)
```

and if `Config` contains such value returns that one, otherwise returns json config, and the default value as a final fallback.

Reply:
`Config` already includes defaults from `JSONConfig`, so this shouldn't be necessary. If `Config` was just `workspace/configuration`, then this function would be useful but we would need to have some sort of way to tell whether something is set or not in each component of `configuration::TopLevel`. For example, if `depth` is set to `0` in `Config`, we would not be able to identify if this is just the default value or a number that is explicitly set. Therefore, we would have to change the type of `depth` in `Eval` to `std::optional<int>`, along with making other fields optional. This would reduce totality throughout the code.